src/equisolve/numpy/models/linear_model.py

```diff
@@ -0,0 +1,257 @@
+                sw_block = sample_weight.block(i_block)
+
+                if len(sw_block.properties) != 1:
+                    raise ValueError(
+                        "Only one property is allowed for sample weights. Given "
+                        f"`sample_weight` contains {len(sw_block.properties)} "
+                        "properties."
+                    )
+
+                if len(X_block.samples) != len(sw_block.samples):
+                    raise ValueError(
+                        f"`X` ({len(X_block.samples)}) and "
+                        f"`sample_weights` ({len(sw_block.samples)}) must "
+                        f"have the same number of samples."
+                    )
+
+    def _numpy_lstsq_solver(self, X, y, sample_weights, alphas):
+        # Convert problem with regularization term into an equivalent
+        # problem without the regularization term
+        num_targets, num_properties = X.shape
+        regularization_all = np.hstack((sample_weights, alphas))
+        regularization_eff = np.diag(np.sqrt(regularization_all))
+        X_eff = regularization_eff @ np.vstack((X, np.eye(num_properties)))
+        y_eff = regularization_eff @ np.hstack((y, np.zeros(num_properties)))
+        least_squares_output = np.linalg.lstsq(X_eff, y_eff, rcond=1e-13)
+        w_solver = least_squares_output[0]
+        return w_solver
+
+    def fit(
+        self, X: TensorMap, y: TensorMap, sample_weight: Optional[TensorMap] = None
+    ) -> None:
+        """Fit Ridge regression model to each block in X.
+
+        :param X: training data
+        :param y: target values
+        :param sample_weight: sample weights
+        """
+        self._validate_data(X, y)
+        self._validate_params(X, sample_weight)
+
+        for i_block, X_block in enumerate(X.blocks()):
+            y_block = y.block(i_block)
+            alpha_block = self.alpha.block(i_block)
+
+            # X_arr has shape of (n_targets, n_properties)
+            X_arr = block_to_array(X_block, self.parameter_keys)
+
+            # y_arr has shape lentgth of n_targets
+            y_arr = block_to_array(y_block, self.parameter_keys)[:, 0]
+
+            # alpha_arr has length of n_properties
+            alpha_arr = alpha_block.values[0]
+
+            # Sample weights
+            if sample_weight is not None:
+                sw_block = sample_weight.block(i_block)
+                # sw_arr has length of n_targets
+                sw_arr = block_to_array(sw_block, self.parameter_keys)[:, 0]
+                assert (
+                    sw_arr.shape == y_arr.shape
+                ), f"shapes = {sw_arr.shape} and {y_arr.shape}"
+            else:
+                sw_arr = np.ones((len(y_arr),))
+
+            w = self._numpy_lstsq_solver(X_arr, y_arr, sw_arr, alpha_arr)
+            self.coef_.append(w.reshape(-1, 1))
+
+        # Convert alpha to a dictionary to be used in external models.
+        self.alpha = tensor_map_to_dict(self.alpha)
```

Comment:
I think this is not a good idea, because if we call the fit method twice it will raise an error the second time. A better way would be to save this in a private variable like `_alpha`. What do you say @agoscinski ?